### PR TITLE
Partial submissions

### PR DESCRIPTION
**Pull request by chiragchhatrala**
## Summary by CodeRabbit

- **New Features**
  - Introduced support for partial form submissions, allowing users to save incomplete forms and resume later.
  - Added a toggle in form settings to enable or disable partial submissions.
  - Enhanced submission interfaces with a new status column and filtering options for “In Progress” versus “Submitted” entries.
  - Updated form statistics to include data on partial submissions for better engagement insights.
  - Implemented a composable function to manage synchronization of partial submissions to the server.
  - Added a dropdown to filter form submissions by status.
  - Integrated partial submission handling throughout form submission logic and UI components.
  - Enhanced form submission settings with a new toggle switch for capturing partial submissions.
  - Introduced handling of submission hashes to support partial submissions.

- **Bug Fixes**
  - Improved validation logic to bypass rules for partial submissions.

- **Documentation**
  - Updated help text for the partial submissions toggle to clarify its purpose.

### CODE REVIEW

File: client/stores/working_form.js [hunk 1]
```diff
@@ -165,7 +166,13 @@ export const useWorkingFormStore = defineStore("working_form", {
       const field = newFields.splice(oldIndex, 1)[0]
       newFields.splice(newIndex, 0, field)
       this.content.properties = newFields
-    }
+    },
+    setSubmissionHash(key, hash) {
```
Comment (JhumanJ) [inline client/stores/working_form.js:170]: We shouldn't store the submission hash here. this is not dependent on currnet form + this is to store form data while we work on it

File: api/database/migrations/2025_02_14_073642_add_partial_submissions_to_form_submissions.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+use Illuminate\Support\Facades\Schema;
+
+return new class () extends Migration {
+    /**
+     * Run the migrations.
+     */
+    public function up(): void
+    {
+        Schema::table('form_submissions', function (Blueprint $table) {
+            $table->enum('status', ['partial', 'completed'])
```
Comment (JhumanJ) [inline api/database/migrations/2025_02_14_073642_add_partial_submissions_to_form_submissions.php:14]: Let's add const to FormSubmission for these 2 values, and use the const instead of the string

File: api/database/migrations/2025_02_14_073642_add_partial_submissions_to_form_submissions.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+                ->default('completed')
+                ->after('form_id')
```
Comment (JhumanJ) [inline api/database/migrations/2025_02_14_073642_add_partial_submissions_to_form_submissions.php:16]: no need for an after()

File: api/database/migrations/2025_02_14_073642_add_partial_submissions_to_form_submissions.php [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+                ->index();
+        });
+
+        Schema::table('forms', function (Blueprint $table) {
+            $table->boolean('enable_partial_submissions')
```
Comment (JhumanJ) [inline api/database/migrations/2025_02_14_073642_add_partial_submissions_to_form_submissions.php:21]: no need for an after

File: api/database/migrations/2025_02_14_073642_add_partial_submissions_to_form_submissions.php (hunk 2 above)
Comment (JhumanJ) [reply api/database/migrations/2025_02_14_073642_add_partial_submissions_to_form_submissions.php:14]: It's there already! Let's use it

File: client/components/open/forms/OpenForm.vue [hunk 5]
```diff
@@ -372,6 +381,9 @@ export default {
 
       this.$refs['form-timer'].stopTimer()
       this.dataForm.completion_time = this.$refs['form-timer'].completionTime
+      if (this.form?.enable_partial_submissions) {
+        this.dataForm.submission_hash = this.workingFormStore.getSubmissionHash(this.pendingSubmission.formPendingSubmissionKey.value)
```
Comment (JhumanJ) [inline client/components/open/forms/OpenForm.vue:385]: Why store it in the store as well? In dataform is enough I believe

File: client/components/open/forms/components/FormStats.vue [hunk 6]
```diff
@@ -118,6 +118,12 @@ export default {
             borderColor: "rgba(16, 185, 129, 1)",
             data: [],
           },
+          {
```
Comment (JhumanJ) [inline client/components/open/forms/components/FormStats.vue:121]: Only if partial submission is enabled

File: client/components/open/forms/components/FormStats.vue [hunk 7]
```diff
@@ -172,6 +178,7 @@ export default {
           this.chartData.labels = Object.keys(statsData.views)
           this.chartData.datasets[0].data = statsData.views
           this.chartData.datasets[1].data = statsData.submissions
+          this.chartData.datasets[2].data = statsData.partial_submissions
```
Comment (JhumanJ) [inline client/components/open/forms/components/FormStats.vue:181]: same only if enabled

File: client/components/open/forms/components/form-components/FormSubmissionSettings.vue [hunk 8]
```diff
@@ -25,6 +25,14 @@
       label="Auto save form response"
       help="Saves form progress, allowing respondents to resume later."
     />
+
+    <ToggleSwitchInput
+      class="mt-4"
+      name="enable_partial_submissions"
+      :form="form"
+      label="Enable partial submissions"
+      help="Allow users to submit incomplete forms. Useful for long forms where you want to analyze drop-off points."
```
Comment (JhumanJ) [inline client/components/open/forms/components/form-components/FormSubmissionSettings.vue:34]: Help text is not claer, users don't actually submit incomplete forms please improve, it's rather the form that keeps track of progress to ensure no data is lost for form owners

File: client/components/open/forms/OpenForm.vue [hunk 9]
```diff
@@ -356,8 +358,15 @@ export default {
       this.isAutoSubmit = true
       this.submitForm()
     }
+    if (!this.adminPreview && this.form?.enable_partial_submissions) {
+      this.partialSubmission.startSync()
```
Comment (JhumanJ) [inline client/components/open/forms/OpenForm.vue:362]: Should only start once user start typing/changing the form content

File: api/app/Models/Forms/FormSubmission.php [hunk 10]
```diff
@@ -9,9 +9,13 @@ class FormSubmission extends Model
 {
     use HasFactory;
 
+    public const STATUS_PARTIAL = 'partial';
+    public const STATUS_COMPLETED = 'completed';
```
Comment (JhumanJ) [inline api/app/Models/Forms/FormSubmission.php:13]: Let's rename `Submitted`

File: client/components/open/forms/components/FormSubmissions.vue [hunk 11]
```diff
@@ -42,6 +42,13 @@
...
           </div>
           <div class="font-semibold flex gap-2">
+            <USelectMenu
+              v-if="form.enable_partial_submissions"
+              v-model="selectedStatus"
+              :options="statusList"
+              value-attribute="value"
+              option-attribute="label"
+            />
             <UButton
```
Comment (JhumanJ) [inline client/components/open/forms/components/FormSubmissions.vue:52]: CleanShot 2025-02-24 at 16 41 59@2x
Display issue - please give fixed with that fits both labels

File: client/components/open/tables/OpenTable.vue [hunk 12]
```diff
@@ -86,6 +97,17 @@
             :value="row[col.id]"
           />
         </td>
+        <td
```
Comment (JhumanJ) [inline client/components/open/tables/OpenTable.vue:100]: Please change label of partial etc, and capitalize. Also make the badge size xs

File: client/composables/forms/usePartialSubmission.js [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+      debouncedSync()
+    }
+  }
+
+  const handleBlur = () => {
+    debouncedSync()
+  }
+
+  const handleBeforeUnload = () => {
+    syncToServer()
```
Comment (JhumanJ) [inline client/composables/forms/usePartialSubmission.js:53]: This doesn't seem to work for me, maybe we need to await? Do we need to display something like an alert message in the meantime? it doesn't send the data before saving

**Review by JhumanJ — CHANGES_REQUESTED** [review]
Hey please check the comments. Also: 

- let's make this a Pro feature 
- every time I reload the form, a new partial submission is created, but since local storage save is enabled it shouldn't. It should only be like this when local storage is not enabled
- files/signature etc don't work. it should upload them and move them the first time it's saved after submisison is created, then after save, dataform should be update to include the file link instead of key. Just like we do when editing existing submissions

thanks!

File: api/app/Http/Controllers/Forms/PublicFormController.php [hunk 14]
```diff
@@ -90,10 +90,38 @@ public function answer(AnswerFormRequest $request, FormSubmissionProcessor $form
...
+        // Remove extra fields from the main data array
+        unset($submissionData['completion_time']);
+        unset($submissionData['submission_hash']);
+
+        // Add submission_id to the submission data if it exists
+        if ($submissionId) {
+            $submissionData['submission_id'] = $submissionId;
+        }
 
         $job = new StoreFormSubmissionJob($form, $submissionData, $completionTime);
```
Comment (JhumanJ) [inline api/app/Http/Controllers/Forms/PublicFormController.php:127]: Below, instead of 

```
else {
            StoreFormSubmissionJob::dispatch($form, $submissionData, $completionTime);
        }
```

we could simply do `$job->dispatch()` or similar to better reafactor

**Review by JhumanJ — CHANGES_REQUESTED** [review]
2 more comments

**Review by JhumanJ — CHANGES_REQUESTED** [review]
- When auto save enabled issue: when refreshing the page and finally submitting, a new record is creating instead of finalizing the pending one (creating duplicates) 
CleanShot 2025-03-06 at 12 39 35@2x
- if I click fill again, then a pending submission is instantly created even though form is still empty
CleanShot 2025-03-06 at 12 42 43@2x
- still. may issues with signatures and files please check again

**Review by JhumanJ — CHANGES_REQUESTED** [review]
Great progress.

**Please add some tests.** 
Examples:
- Test Edge Cases: Ensure automated tests cover scenarios like:
- Submitting partially, then completing using the returned hash.
- Submitting partially multiple times for the same submission.
- Attempting partial submission when the feature is disabled for the form.
- Correct stats calculation for partial vs. completed submissions.
- File/signature handling during partial saves.

**Check database feature clash**
Also there's a feature clash with "does not exist in submission" or "exisits in submisison". we when checking for these, we need to filter out pending submission otherwise pending submission conflicts with itself.

**Review by JhumanJ — APPROVED** [review]

**Review by JhumanJ — APPROVED** [review]
